### ISSUE TITLE
Improve diagnostics for failed call to overloaded function

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#274]
Currently we emit a very opaque "No overload of bound method `__init__` matches" diagnostic, with zero additional context: no reference to the called function, its overloads, what failed to match, etc.

Sample code:

```py
from typing import overload

@overload
def f(x: int) -> int: ...

@overload
def f(x: str) -> str: ...

def f(x: int | str) -> int | str:
    return x

f(b"foo")
```

**Comment by sharkdp** [comment]
This has been closed in https://github.com/astral-sh/ruff/pull/18073

**Comment by dhruvmanila** [comment]
Re-opening this because this currently only works for `FunctionType`, it should be applied to other overloaded callables as well like:
* `BoundMethodType` - contains the `FunctionType`

  ```py
  import tempfile
  
  tempfile.TemporaryDirectory()
  ```

* `ClassLiteral` - this will be tricky because it could be either `__new__` or `__init__`

   ```py
   type()
   ```

Others are mentioned [here](https://github.com/astral-sh/ruff/blob/466021d5e1793ca30e0d860cb1cd27e32f5233aa/crates/ty_python_semantic/src/types/call/bind.rs#L1504-L1523) for which I'm not sure if it's worth the effort.

cc @BurntSushi, feel free to close it if you prefer to track this in a separate issue

**Comment by BurntSushi** [comment]
I created #409 to track remaining work here. Thanks for the call out @dhruvmanila!
